src/main/java/org/folio/rest/transaction/PaymentCreditHandler.java

```diff
@@ -0,0 +1,373 @@
+      .getCurrency());
+    tempPayments.forEach(pymtTxn -> {
+      Transaction encumbranceTxn = encumbrancesMap.get(pymtTxn.getPaymentEncumbranceId());
+      double newAwaitingPayment = MoneyUtils.subtractMoney(encumbranceTxn.getEncumbrance()
+        .getAmountAwaitingPayment(), pymtTxn.getAmount(), currency);
+      double newExpended = MoneyUtils.sumMoney(encumbranceTxn.getEncumbrance()
+        .getAmountExpended(), pymtTxn.getAmount(), currency);
+      double newAmount = MoneyUtils.subtractMoneyNonNegative(encumbranceTxn.getAmount(), pymtTxn.getAmount(), currency);
+
+      encumbranceTxn.getEncumbrance()
+        .setAmountAwaitingPayment(newAwaitingPayment);
+      encumbranceTxn.getEncumbrance()
+        .setAmountExpended(newExpended);
+      encumbranceTxn.setAmount(newAmount);
+    });
+
+    return encumbrancesMap;
+  }
+
+  private Future<List<Transaction>> getAllEncumbrances(Tx<List<Transaction>> tx) {
+    Promise<List<Transaction>> promise = Promise.promise();
+    String sql = "SELECT DISTINCT ON (permtransactions.id) permtransactions.jsonb " + "FROM "
+        + getFullTableName(getTenantId(), TRANSACTIONS_TABLE) + " AS permtransactions " + "INNER JOIN "
+        + getFullTemporaryTransactionTableName() + " AS transactions "
+        + "ON transactions.paymentEncumbranceId = permtransactions.id" + " WHERE transactions.sourceInvoiceId = ?";
+    JsonArray params = new JsonArray();
+    params.add(getSummaryId(tx.getEntity()
+      .get(0)));
+    tx.getPgClient()
+      .select(tx.getConnection(), sql, params, reply -> {
+        if (reply.failed()) {
+          handleFailure(promise, reply);
+        } else {
+          List<Transaction> encumbrances = reply.result()
+            .getResults()
+            .stream()
+            .flatMap(JsonArray::stream)
+            .map(o -> new JsonObject(o.toString()).mapTo(Transaction.class))
+            .collect(Collectors.toList());
+          promise.complete(encumbrances);
+        }
+      });
+    return promise.future();
+  }
+
+
+
+  @Override
+  public void updateTransaction(Transaction transaction) {
+    throw new UnsupportedOperationException("Payments and credits are Immutable");
+  }
+
+  @Override
+  void handleValidationError(Transaction transaction) {
+
+    List<Error> errors = new ArrayList<>();
+
+    if (isCreditTransaction.test(transaction)) {
+      errors.addAll(buildNullValidationError(transaction.getToFundId(), "toFundId"));
+    } else {
+      errors.addAll(buildNullValidationError(transaction.getFromFundId(), "fromFundId"));
+    }
+    if (isNotEmpty(errors)) {
+      throw new HttpStatusException(422, JsonObject.mapFrom(new Errors().withErrors(errors)
+        .withTotalRecords(errors.size()))
+        .encode());
+    }
+
+  }
+
+  @Override
+  String createTempTransactionQuery() {
+    return "INSERT INTO " + getFullTemporaryTransactionTableName() + " (id, jsonb) VALUES (?, ?::JSON) ";
```

Comment:
In order to be able to retry the payment of the invoice after the error, the create temporary transaction has to be the idempotent.

Reply:
added ON CONFLICT condition, and a test case